### ISSUE TITLE
`classes` + `max_det` silently drops valid detections on end-to-end models (YOLO26/YOLOv10 and NMS-embedded exports)

### ISSUE THREAD

**Issue by JESUSROYETH** [ultralytics/ultralytics#25044]
### Bug

Combining `classes` with `max_det` on end-to-end models (YOLO26, YOLOv10, and any model exported with embedded NMS) silently returns fewer detections than the model found, down to zero, while the same arguments on YOLO11 work as expected. There is no warning, so migrating a pipeline from YOLO11 to YOLO26 changes results with identical code.

```python
from ultralytics import YOLO

img = "https://ultralytics.com/images/bus.jpg"
for k in (300, 4, 1):
    print(k, len(YOLO("yolo26n.pt")(img, classes=[0], max_det=k, verbose=False)[0].boxes))
print("yolo11n", len(YOLO("yolo11n.pt")(img, classes=[0], max_det=1, verbose=False)[0].boxes))
```

Output on 8.4.90 (main @ e821386, also reproduced on 8.4.87):

| Call (bus.jpg) | Boxes returned |
|---|---|
| yolo26n.pt `classes=[0], max_det=300` | 4 persons (0.913 / 0.905 / 0.870 / 0.556) |
| yolo26n.pt `classes=[0], max_det=4` | 3 persons |
| yolo26n.pt `classes=[0], max_det=1` | **0 boxes** |
| yolo26n.pt no `classes`, `max_det=1` | 1 (bus, 0.924) |
| yolo11n.pt `classes=[0], max_det=1` | 1 person (0.888) |

With `max_det=1` the user asks for the best person, the image has four persons above 0.55 conf, and gets an empty result. Exported models hit the same path: yolo11n.onnx exported with `nms=True` and yolo26n.onnx (native end2end) both return 0 boxes with `classes=[0], max_det=1`.

### Root cause

The `max_det` truncation runs before the `classes` filter, in two stages:

1. The end-to-end branch of `non_max_suppression` (`ultralytics/utils/nms.py:67`) slices `[:max_det]` first and applies `classes` after. With `max_det=1` the slice keeps only the bus, and the classes filter deletes it.
2. For torch end2end models the head top-k is also lowered to `args.max_det` in `setup_model` (`ultralytics/engine/predictor.py:402`, since #23396), so with `max_det=1` the head already returned a single candidate before nms.py sees anything.

The NMS-based branch of the same function filters `classes` first (`nms.py:126-131`) and truncates at the end (`nms.py:157`), and `RTDETRPredictor.postprocess` does the same (`models/rtdetr/predict.py:62-64`) — the end-to-end branch is the only place with the inverted order. Historically, #13971 (v8.2.43) added the `classes` filter to this branch with no truncation; the `[:max_det]` slice came with #15917 (v8.2.85) and landed between the conf filter and the classes filter.

A related observation, out of scope here: `setup_model` runs once per predictor, so the `max_det` of the first predict call stays applied to the head of a reused model instance even without `classes`.

### Environment

- Ultralytics 8.4.90 (main @ e821386) and 8.4.87
- Python 3.12, torch 2.12.1+cu130, RTX 4060, Ubuntu

I am opening a PR with a fix for both stages (reorder the end-to-end branch to filter before truncating, and keep the head top-k at its 300 default so the filter sees all candidates), verified against a 48-case matrix with bit-identical results outside the buggy cases.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JESUSROYETH, thank you for the detailed bug report and for your interest in Ultralytics 🚀! We appreciate the clear reproduction snippet, environment details, root-cause notes, and mention of an incoming PR. This is an automated response, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please ensure the provided reproduction remains a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) that maintainers can run directly to help us debug and validate the behavior.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Thanks for debugging and opening a PR.
